rlkit/core/rl_algorithm.py

```diff
@@ -482,6 +495,128 @@ def get_extra_data_to_save(self, epoch):
+            final_returns.append(all_rets[-1])
+            online_returns.append(all_rets)
+        return final_returns, online_returns
+
+    def evaluate(self, epoch):
+        statistics = OrderedDict()
+        statistics.update(self.eval_statistics)
+        self.eval_statistics = statistics
+
+        ### sample trajectories from prior for debugging / visualization
+        if self.dump_eval_paths:
+            prior_paths = []
+            for _ in range(100 // self.num_steps_per_eval):
+                # just want stochasticity of z, not the policy
+                prior_paths += self.eval_sampler.obtain_samples(deterministic=True, resample=np.inf)
+            logger.save_extra_data(prior_paths, path='eval_trajectories/prior-epoch{}'.format(epoch))
+
+        ### train tasks
+        # eval on a subset of train tasks for speed
+        indices = np.random.choice(self.train_tasks, len(self.eval_tasks))
+        dprint('evaluating on {} train tasks'.format(len(indices)))
+        ### eval train tasks with posterior sampled from the training replay buffer
+        train_returns = []
+        for idx in indices:
+            self.task_idx = idx
+            self.env.reset_task(idx)
+            paths = []
+            # TODO: put these lines into method evaluate_task(idx) that goes into sac.py?
+            for _ in range(10):
+                self.infer_posterior(idx)
+                paths += self.eval_sampler.obtain_samples(num_samples=self.max_path_length + 1, deterministic=True, resample=np.inf)
+            if self.sparse_rewards:
+                for p in paths:
+                    sparse_rewards = np.stack(e['sparse_reward'] for e in p['env_infos']).reshape(-1, 1)
+                    p['rewards'] = sparse_rewards
+            train_returns.append(eval_util.get_average_returns(paths))
+        train_returns = np.mean(train_returns)
+        ### eval train tasks with on-policy data to match eval of test tasks
+        train_final_returns, train_online_returns = self._do_eval(indices, epoch)
+        print('train online returns')
+        print(train_online_returns)
+
+        ### test tasks
+        # TOOD: should this be using the dprint in eval_util or pythons own dprint?
+        dprint('evaluating on {} test tasks'.format(len(self.eval_tasks)))
+        test_final_returns, test_online_returns = self._do_eval(self.eval_tasks, epoch)
+        print('test online returns')
+        print(test_online_returns)
+
+        # ewww, this section breaks abstraction
+        # should we just move eval into sac?
+        # save the final posterior
+        if self.use_information_bottleneck:
+            z_mean = np.mean(np.abs(ptu.get_numpy(self.policy.z_means[0])))
+            z_sig = np.mean(ptu.get_numpy(self.policy.z_vars[0]))
+            self.eval_statistics['Z mean eval'] = z_mean
+            self.eval_statistics['Z variance eval'] = z_sig
+
+        # TODO(KR) what does this do
```

Comment:
I don't think any of our envs support this and it doesn't seem like it's needed?

Reply:
We don't use it now, but it could be useful for logging env specific things like final distance to goal. Probably not that useful for metalearning given all the different envs, but I can see it being useful in the POMDP setting.